File: client/src-tauri/src/model/task.rs
use crate::model::time_stamps::Timestamps;
use serde::{Deserialize, Serialize};

use crate::{
    db::table::HasKey,
    model::{macro_model::model_with_default, default_value::DefaultValue},
};

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
#[serde(rename_all = "PascalCase")]
pub enum TaskStatus {
    NotStarted,
    InProgress,
    Done,
}

impl DefaultValue for TaskStatus {
    fn default_value() -> Self {
        TaskStatus::NotStarted
    }
}

impl HasKey<(String, String)> for Task {
    fn key(&self) -> (String, String) {
        (self.project_id.clone(), self.id.clone())
    }
}

model_with_default!(
    Task,
    TaskRequest,
    {
        id: String,
        project_id: String,
    },
    {
        phase_id: String,
        name: String,
        planned_start: Option<String>,
        planned_end: Option<String>,
        planned_hours: Option<f64>,
        actual_start: Option<String>,
        actual_end: Option<String>,
        actual_hours: Option<f64>,
        progress_rate: f64,
        status: TaskStatus,
    }
);

File: client/src/types/db/task.ts
import { Timestamps } from "./common";

export enum TaskStatus {
    NotStarted = "NotStarted",
    InProgress = "InProgress",
    Done = "Done",
}

export type TaskKey = {
    id: string,
    project_id: string,
}

export type TaskValue = {
    phase_id: string,
    name: string,

    planned_start: string,
    planned_end: string,
    planned_hours: number,

    actual_start: string,
    actual_end: string,
    actual_hours: number,

    progress_rate: number,
    status: TaskStatus,
    timestamps: Timestamps,
}

// Row = Key + Value
export type TaskRow = TaskKey & TaskValue;
export type TaskPayload = TaskKey & Partial<TaskValue>;

export const toTaskPayload = (param: TaskRow): TaskPayload => (
    {
        id: param.id,
        project_id: param.project_id,
        phase_id: param.phase_id,
        name: param.name,
        planned_start: param.planned_start,
        planned_end: param.planned_end,
        planned_hours: param.planned_hours,
        actual_start: param.actual_start,
        actual_end: param.actual_end,
        actual_hours: param.actual_hours,
        progress_rate: param.progress_rate,
        status: param.status,
    }
)
